README.md
# specgram

![build](https://github.com/rimio/specgram/workflows/build/badge.svg)

Small program that computes and plots spectrograms, either in a live window or to disk, with support for stdin input.

## Preview

![example](resources/readme_images/example.png?raw=true "Example")

## Build and install

If you are using ArchLinux you can grab the latest release from the AUR package [specgram](https://aur.archlinux.org/packages/specgram/), or get the main branch with [specgram-git](https://aur.archlinux.org/packages/specgram-git/).

Otherwise, you can build and install the program from source:
```bash
# Clone the repo
git clone https://github.com/rimio/specgram.git
cd specgram && mkdir build && cd build

# Build
cmake ..
make

# Install
sudo make install
```

## Dependencies

This program dynamically links against [FFTW](http://www.fftw.org/) and [SFML 2.5](https://www.sfml-dev.org/).

The source code of [Taywee/args](https://github.com/Taywee/args) is embedded in the program (see ```src/args.hxx```).

## Usage

For a complete description of the program functionality please see the [manpage](man/specgram.1.pdf).

### Input and output modes

**specgram** has two mutually exclusive input methods: from standard input and from file.

In order to generate a spectrogram from an input file ```infile``` and write the output to ```output.png```:

```bash
specgram -i infile outfile.png
```

If no input file is specified, then the default behaviour is to read input data *indefinitely* from standard input.
For example, we can query PulseAudio for audio sources:

```bash
$ pactl list sources short
1	alsa_output.usb-BEHRINGER_UMC204HD_192k-00.analog-surround-40.monitor	module-alsa-card.c	s16le 4ch 44100Hz	IDLE
2	alsa_input.usb-BEHRINGER_UMC204HD_192k-00.analog-stereo	module-alsa-card.c	s32le 2ch 44100Hz	SUSPENDED
3	alsa_output.pci-0000_00_1f.3.iec958-stereo.monitor	module-alsa-card.c	s16le 2ch 44100Hz	SUSPENDED
4	stereo.A.monitor	module-remap-sink.c	s16le 2ch 44100Hz	IDLE
5	stereo.B.monitor	module-remap-sink.c	s16le 2ch 44100Hz	SUSPENDED
11	alsa_output.pci-0000_01_00.1.hdmi-stereo.monitor	module-alsa-card.c	s16le 2ch 44100Hz	SUSPENDED

$ export PASOURCE="stereo.A.monitor"
```

In my case the default sink is ```stereo.A```, and I use the monitor source ```stereo.A.monitor``` to capture what I'm hearing in my headphones:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram outfile.png
[2020-12-27 15:56:15.058] [info] Creating 1024-wide FFTW plan
[2020-12-27 15:56:15.058] [info] Input stream: signed 16bit integer at 44100Hz
```

The program will keep reading data and computing FFTs until end of times or until it receives a ```SIGINT```.
In the Linux terminal this can be achieved by pressing ```CTRL+C```.

Once the signal is received, the program stops reading data from input and writes to ```outfile.png``` whatever it cached so far:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram outfile.png
[2020-12-27 15:57:29.967] [info] Creating 1024-wide FFTW plan
[2020-12-27 15:57:29.967] [info] Input stream: signed 16bit integer at 44100Hz
^C[2020-12-27 15:57:31.813] [info] Terminating ...
```

It is sometimes useful to see the spectrogram in real time. Live mode can be enabled with the ```-l, --live``` flag:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -l outfile.png
```

The spectrogram will now be displayed as it is being compute from standard input.
When either ```SIGINT``` is received or the live window is closed, the program will terminate and write ```outfile.png```.

If file output is not desired, only live mode can be used, and nothing is written upon termination:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -l
```

For obvious reasons, live mode cannot be used with file input.

### Input options

In the above examples we assumed that the program input is 16-bit signed integer at 44.1kHz, which happens to be what my sound card (and many others) outputs by default.

We can, however, specify any other rate with ```-r, --rate``` and datatype with ```-d, --datatype```:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw --format=float32 --rate=48000 | specgram -l -r 48000 -d f32
```

The above example will read 32-bit floating point input at 48kHz.
For a full list of supported data types see the [manpage](man/specgram.1.pdf).

**NOTE:** The specified rate is used only for display purposes and for interpreting other command line parameters.
There's nothing stopping us from using a different rate than the actual device rate, going down to nanohertz:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -l -r 1e-8
```

### FFT options

The FFT window width can be specified with ```-f, --fft_width``` and the stride, that is the distance between the beginning of two subsequent FFT windows, can be specified with ```-g, --fft_stride```:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -l -f 2048 -g 1024 
```

The above will compute 2048 elements wide FFT windows with a stride of 1024 elements, that is with a 50% overlap between windows.

Usually, a larger FFT window will give better frequency resolution but worse time resolution (i.e. it will be harder to locate signals in the time domain).

A smaller stride will give a *smoother* and richer output, but will strain the CPU more.

**NOTE:** You will notice that there isn't much difference between the output of the above command and the others.
That is because the display width is different from the FFT window width.
To change the display width, see **Display Options** below.

Lastly, if you encounter high sample rate signals, for which you can't display a wide enough (or often enough) window, you can use window averaging (```-A, --average```).

```bash
$ rx_sdr -d 0 -g 50 -f 97300000 -s 960000 -F CF32 - | ./specgram -lq -r 960000 -d cf32 -A 20 
```

The above example consumes input at 960k samples per second from a RTL-SDR dongle, which at a 1024 wide FFT window would mean displaying over 900 windows per second; a bit much for the average PC, and for the average human to follow.

Averaging 20 windows gives us a much more reasonable 47 windows per second.

### Display options

To change the display width we can use ```-w, --width```:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -l -f 1024 -w 1024
```

As you will notice, the spectrogram is somewhat blurry, because the program is resampling the 513 element wide positive part of the FFT output to the display width of 1024.
If you need sharp, crisp spectrograms, then you can use ```-q, --no_resampling``` to disable resampling:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -lq -f 1024
```
When not resampling, you can no longer specify the display width, as it is computed from the rest of the parameters.

Another use case is displaying a specific band of frequencies, using ```-x, --fmin``` and ```-y, --fmax``` to set the frequency bounds.
For example, to zoom in on the 500-3000Hz band:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -lq -f 9192 -x 500 -y 3000 
```

The colormap can be specified with ```-c, --colormap```; see image below the example for supported colormaps:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -l -c orange
```

![colormaps](resources/readme_images/colormaps.png?raw=true "Colormaps")

In order to see how to specify the background, foreground and custom colormap colors, please see the [manpage](man/specgram.1.pdf).

While the live view has both axes and legend enabled by default, file output does not.
To enable them use ```-a, --axes``` or ```-e, --legend```.
Please note that axes are implicit when displaying the legend.

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -le outfile.png
```

Finally, if you'd like to rotate the spectrogram 90 degrees counter-clockwise, so as to read it from left to right, you can use ```-z, --horizontal```:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -lz
```

This flag applies to both the live window and output file.

### Live options

Use ```-k, --count``` to control the number of FFT windows displayed in live view:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -l -k 128
```

Use ```-t, --title``` to specify the live window title:

```bash
$ parec --channels=1 --device="${PASOURCE}" --raw | specgram -l -t "My spectrogram"
```

## License

Copyright (c) 2020-2023 Vasile Vilvoiu \<vasi@vilvoiu.ro>

**specgram** is free software; you can redistribute it and/or modify it under the terms of the MIT license.
See LICENSE for details.

## Acknowledgements

Taywee/args library by Taylor C. Richberger and Pavel Belikov, released under the MIT license.

Program icon by Flavia Fabian, released under the CC-BY-SA 4.0 license.

Share Tech Mono font by Carrois Type Design, released under Open Font License.

Special thanks to Eugen Stoianovici for code review and various fixes.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.9.3] - 2023-05-06
### Added
- Support to exit program with `Esc` key.
- Support to wait for input with `-S` or `--sleep_for_input`; program can now _not_ busywait on sparse input.
- Two new colormaps, `hot` and `inferno`; the latter is the new default.
- Support for in-source build.

### Fixed
- Usage of bad version macros in generated header file.
- Various warnings and compilation issues on newer versions of `g++`.

## [0.9.2] - 2021-07-22
### Added
- Unit tests.
- Header file comments.

### Removed
- Removed `spdlog` as a dependency due to the `fmt 8.0.0` debacle.

### Fixed
- \[SERIOUS\] Window function term computation ([commit](https://github.com/rimio/specgram/commit/46f0cc9395ffc841626d9242238868383c146233)).
- Axes ticks computation is now much more refined.
- Assertion fail when frequency band and window size are very small.

## [0.9.1] - 2021-07-16
### Added
- Support for `-` as input or output filename, indicating `stdin`/`stdout`.
- Ability to dump the PNG encoded version of output to `stdout`.
- Support for arbitrary scales with custom units.
- Support for linear scales (as opposed to the logarithmic dBFS scale supported in 0.9.0).

### Changed
- Main window is rendered on each iteration of the main loop. Underlying texture is still only updated when new windows are computed.

### Fixed
- \[SERIOUS\] Branch on uninitialized variable ([commit](https://github.com/rimio/specgram/commit/b13609afcdf66d781db70fb75f6869a052a49079)).
- Garbled window texture when no input is received.
- Tick labels no longer overlap; enforce minimum tick spacing based on text width. 


src/specgram.cpp
/*
 * Copyright (c) 2020-2023 Vasile Vilvoiu <vasi@vilvoiu.ro>
 *
 * specgram is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */
#include "configuration.hpp"
#include "input-parser.hpp"
#include "input-reader.hpp"
#include "color-map.hpp"
#include "value-map.hpp"
#include "window-function.hpp"
#include "fft.hpp"
#include "live.hpp"

#include <iostream>
#include <iomanip>
#include <fstream>
#include <csignal>
#include <list>
#include <random>
#include <cstdio>
#include <cassert>
#include <thread>
#include <chrono>

/* main loop exit condition */
volatile bool main_loop_running = true;

/* temporary output file name */
std::string temp_file_name = "";

/*
 * logger - logging is minimal and only happens in this file
 */
#define RESET "\033[0m"
#define GREEN "\033[32m"
#define YELLOW "\033[33m"
#define RED "\033[31m"

#define INFO(str) { std::cerr << "[" << GREEN << "INFO" << RESET << "] " << str << std::endl; }
#define WARN(str) { std::cerr << "[" << YELLOW << "WARN" << RESET << "] " << str << std::endl; }
#define ERROR(str) { std::cerr << "[" << RED << "ERROR" << RESET << "] " << str << std::endl; }

/*
 * SIGINT handler
 */
void
sigint_handler(int)
{
    /* no longer loop */
    main_loop_running = false;

    /* uninstall handler in case user REALLY does not want to wait for wrap-up */
    std::signal(SIGINT, nullptr);
}

/*
 * printing functions
 */
void
print_complex_window(const std::string& name, const ComplexWindow& window)
{
    std::cout << name << ": [";
    for (const auto& v : window) {
        std::cout << " " << std::setprecision(3) << std::fixed << v.real()
                  << std::showpos << v.imag() << "j";
    }
    std::cout << "]" << std::endl;
}

void
print_real_window(const std::string& name, const RealWindow& window)
{
    std::cout << name << ": [";
    for (const auto& v : window) {
        std::cout << " " << std::setprecision(3) << std::fixed << v;
    }
    std::cout << "]" << std::endl;
}

/*
 * generate a random string (see https://stackoverflow.com/a/50556436)
 */
std::string
generate_random_string(std::size_t length)
{
    std::mt19937 generator { std::random_device{}() };
    std::uniform_int_distribution<int> distribution { 'a', 'z' };

    std::string output(length, '\0');
    for(auto& ch : output) {
        ch = static_cast<char>(distribution(generator) & 0xff);
    }

    return output;
}

/*
 * dump image to stdout (in PNG format). This writes a temp file in /dev/shm since SFML cannot save to memory.
 */
void
dump_to_stdout(const sf::Image& image)
{
    static constexpr size_t TEMP_BUFFER_SIZE = 1024;
    static constexpr size_t TEMP_FILENAME_LENGTH = 32;

    /* save */
    temp_file_name = "/dev/shm/" + generate_random_string(TEMP_FILENAME_LENGTH) + ".png";
    INFO("Temporary file: " << temp_file_name);
    image.saveToFile(temp_file_name);

    /* from now on we have a leakable resource (the file); if using STDIN for input, we're here from a SIGINT,
     * and we expect a SIGPIPE soon; install a handler that will clean up */
    std::signal(SIGPIPE, [](int) { std::remove(temp_file_name.c_str()); /* no logger, no checks */ std::exit(0); });

    /* dump */
    std::ifstream file(temp_file_name, std::ios::in | std::ios::binary);
    if (file.fail()) {
        throw std::runtime_error("cannot read temp file " + temp_file_name);
    }
    while (!file.eof()) {
        char buffer[TEMP_BUFFER_SIZE];
        file.read(buffer, TEMP_BUFFER_SIZE);
        std::size_t read_count = file.gcount();
        std::cout.write(buffer, read_count);
    }
    file.close();

    /* clean up */
    if (std::remove(temp_file_name.c_str()) != 0) {
        WARN("Failed to delete temp file " << temp_file_name);
    }
}

/*
 * entry point
 */
int
main(int argc, char** argv)
{
    /* parse command line arguments into global settings */
    auto [conf, conf_rc, conf_must_exit] = Configuration::Build(argc, (const char **)argv);
    if (conf_must_exit) {
        return conf_rc;
    }

    /* decide whether we have output or not */
    bool have_output = conf.GetOutputFilename().has_value() || conf.MustDumpToStdout();

    /* create window function */
    auto win_function = WindowFunction::Build(conf.GetWindowFunction(), conf.GetFFTWidth());

    /* create FFT */
    INFO("Creating " << conf.GetFFTWidth() << "-wide FFTW plan");
    FFT fft(conf.GetFFTWidth(), win_function);

    /* create value map */
    INFO("Scale " << (conf.GetScaleType() == ValueMapType::kLinear ? "linear" : "decibel") <<
         ", unit " << conf.GetScaleUnit() << ", bounds [" << conf.GetScaleLowerBound() <<
         ", " << conf.GetScaleUpperBound() << "]");
    std::unique_ptr<ValueMap> value_map = ValueMap::Build(conf.GetScaleType(),
                                                          conf.GetScaleLowerBound(),
                                                          conf.GetScaleUpperBound(),
                                                          conf.GetScaleUnit());

    /* create color map */
    auto color_map = ColorMap::Build(conf.GetColorMap(), conf.GetBackgroundColor(),
                                     conf.GetColorMapCustomColor());

    /* create live window */
    std::unique_ptr<LiveOutput> live = nullptr;
    if (conf.IsLive()) {
        live = std::make_unique<LiveOutput>(conf);
        live->Render(); /* render empty window */
    }

    /* create input parser */
    auto input = InputParser::Build(conf.GetDataType(), conf.GetPrescaleFactor(), conf.HasComplexInput());
    if (input == nullptr) {
        return 1;
    }

    /* create input reader */
    std::istream *input_stream = nullptr;
    std::unique_ptr<InputReader> reader = nullptr;
    if (conf.GetInputFilename().has_value()) {
        INFO("Input: " << *conf.GetInputFilename());
        input_stream = new std::ifstream(*conf.GetInputFilename(), std::ios::in | std::ios::binary);
        assert(input_stream != nullptr);
        if (!input_stream->good()) {
            ERROR("Failed to open input file " << *conf.GetInputFilename());
            return 1;
        }
        reader = std::make_unique<SyncInputReader>(input_stream,
                                                   input->GetDataTypeSize() * conf.GetBlockSize());
    } else {
        INFO("Input: STDIN");
        input_stream = &std::cin;
        reader = std::make_unique<AsyncInputReader>(input_stream,
                                                    input->GetDataTypeSize() * conf.GetBlockSize());
    }

    /* display initialization info */
    if (input->IsFloatingPoint()) {
        INFO("Input stream: " << (input->IsComplex() ? "complex " : "") << input->GetDataTypeSize() * 8 <<
             "bit floating point at " << conf.GetRate() << "Hz");
    } else {
        INFO("Input stream: " << (input->IsComplex() ? "complex " : "") <<
             (input->IsSigned() ? "signed " : "unsigned ") << input->GetDataTypeSize() * 8 <<
             "bit integer at " << conf.GetRate() << "Hz");
    }

    /* install SIGINT handler for CTRL+C */
    std::signal(SIGINT, sigint_handler);

    /* FFT window history */
    std::list<std::vector<uint8_t>> history;

    /* window average */
    RealWindow window_sum;
    window_sum.resize(conf.GetWidth());
    std::size_t window_sum_count = 0;

    /* main loop */
    while (main_loop_running && !reader->ReachedEOF()) {
        /* check for window events (if necessary) and redraw */
        if (live != nullptr) {
            if (!live->HandleEvents()) {
                /* exited by closing window */
                main_loop_running = false;
                /* uninstall signal so that reader thread can exit successfully */
                std::signal(SIGINT, nullptr);
            }
            live->Render();
        }

        /* check for a complete block */
        auto block = reader->GetBlock();
        if (!block) {
            /* block not finished yet */
            if (auto sleep = conf.GetSleepForInput()) {
                /* sleep for a bit so we don't busywait on sparse input */
                std::this_thread::sleep_for(std::chrono::duration<size_t, std::milli>(sleep));
            }
            continue;
        }

        /* take whatever is available from input stream */
        auto pvc = input->ParseBlock(*block);
        assert(pvc == block->size() / input->GetDataTypeSize());

        /* check if we have enough for a new FFT window */
        if ((input->GetBufferedValueCount() < conf.GetFFTWidth())
            || (input->GetBufferedValueCount() < conf.GetFFTStride())) {
            /* wait until we get enough values for a window and the spacing between windows */
            continue;
        }

        /* retrieve window and remove values that won't be used further */
        auto window_values = input->PeekValues(conf.GetFFTWidth());
        input->RemoveValues(conf.GetFFTStride());
        if (conf.MustPrintInput()) {
            print_complex_window("input", window_values);
        }

        /* compute FFT on fetched window */
        auto fft_values = fft.Compute(window_values);
        if (conf.MustPrintFFT()) {
            print_complex_window("fft", fft_values);
        }

        /* compute magnitude */
        auto fft_magnitude = FFT::GetMagnitude(fft_values, conf.IsAliasingNegativeFrequencies());

        /* map magnitude to [0..1] domain */
        auto normalized_magnitude = value_map->Map(fft_magnitude);

        if (conf.CanResample()) {
            /* resample to display width */
            normalized_magnitude = FFT::Resample(normalized_magnitude, conf.GetRate(), conf.GetWidth(),
                                                 conf.GetMinFreq(), conf.GetMaxFreq());
        } else {
            /* crop to display width */
            normalized_magnitude = FFT::Crop(normalized_magnitude, conf.GetRate(),
                                             conf.GetMinFreq(), conf.GetMaxFreq());
        }
        if (conf.MustPrintOutput()) {
            print_real_window("output", normalized_magnitude);
        }

        /* add to running total */
        assert(window_sum.size() == normalized_magnitude.size());
        assert(conf.GetAverageCount() > 0);
        for (std::size_t i = 0; i < window_sum.size(); i++) {
            window_sum[i] += normalized_magnitude[i] / (double)conf.GetAverageCount();
        }
        window_sum_count++;

        if (window_sum_count < conf.GetAverageCount()) {
            /* we still have to compute some more windows before we show */
            continue;
        }

        /* add to live */
        if (live != nullptr) {
            auto colorized = live->AddWindow(window_sum);
            if (have_output) {
                history.push_back(colorized);
            }
        } else if (have_output) {
            history.push_back(color_map->Map(window_sum));
        }

        /* reset */
        window_sum_count = 0;
        for (auto& v : window_sum) {
            v = 0.0f;
        }
    }
    INFO("Terminating ...");

    /* close input file */
    if (conf.GetInputFilename().has_value()) {
        assert(input_stream != nullptr);
        assert(input_stream != &std::cin);
        delete input_stream;
        input_stream = nullptr;
    }

    /* save file */
    if (have_output) {
        Renderer file_renderer(conf, history.size());
        file_renderer.RenderFFTArea(history);
        auto image = file_renderer.GetCanvas().copyToImage();

        /* rotate, if needed */
        if (conf.IsHorizontal()) {
            std::size_t w = image.getSize().x;
            std::size_t h = image.getSize().y;

            /* allocate memory */
            auto iptr = reinterpret_cast<const uint32_t *>(image.getPixelsPtr());
            uint32_t *optr = new uint32_t[w * h];

            /* copy 4 bytes (one pixel) at a time */
            for (std::size_t l = 0; l < h; l ++) {
                auto in = iptr + l * w;
                auto out = optr + (w-1) * h + l;
                for (std::size_t c = 0; c < w; c++, in++, out -= h) {
                    *out = *in;
                }
            }

            /* create rotated image */
            sf::Image rimage;
            rimage.create(h, w, reinterpret_cast<const uint8_t *>(optr));
            delete[] optr;

            image = rimage;
        }

        /* dump to file or stdout */
        if (conf.GetOutputFilename().has_value()) {
            INFO("Output: " << *conf.GetOutputFilename());
            image.saveToFile(*conf.GetOutputFilename());
        } else if (conf.MustDumpToStdout()) {
            INFO("Output: STDOUT");
            dump_to_stdout(image);
        } else {
            throw std::runtime_error("don't know what to do with output");
        }
    }

    /* all ok */
    return 0;
}
